p2p/host/basic/basic_host.go

```diff
@@ -723,8 +723,10 @@ func (h *BasicHost) Connect(ctx context.Context, pi peer.AddrInfo) error {
 	h.Peerstore().AddAddrs(pi.ID, pi.Addrs, peerstore.TempAddrTTL)
 
 	forceDirect, _ := network.GetForceDirectDial(ctx)
+	canUseTransient, _ := network.GetUseTransient(ctx)
 	if !forceDirect {
-		if h.Network().Connectedness(pi.ID) == network.Connected {
+		connectedness := rh.Network().Connectedness(pi.ID)
```

Comment:
I'm not sure if this is now correct, but it was definitely wrong before. We need to _block_ until we have a direct connection. Does that now work?

Reply:
There is inconsistency between Connect and NewStream. 

Connect will just connect and doesn't care about whether the connection is transient or not. Using ForceDirect tells connect to *only* dial direct addresses and not relayed addresses. This is used by the DCUTR code to ensure that relayed addresses are not dialled. 

NewStream cares about whether the connection is transient or not and will block for hole punching. So NewStream uses Connect to first connect to the peer over a relay and then waits for the hole punch to succeed if WithUseTransient option is not selected. 

The fix here is to have NewStream call Connect with network.WithUseTransient and then wait for the hole punch to happen. 

I suggest we keep it as it is in this PR and take that change in a separate PR.

Reply:
I disagree. The fix is to make `Connect` wait until you have a connection with the desired properties, not to mess with `NewStream`. Wasn't that fixed in https://github.com/libp2p/go-libp2p/issues/1603?

Reply:
That fixed it for NewStream which now blocks for a direct connection and kept Connect behaviour the same as before. Part of the reason was because Connect was broken previously. Connect without network.WithUseTransient would return with no error and transient connectivity without blocking.  

Thinking more about this my suggested solution is incorrect and I'm not against fixing Connect behaviour. Just to take it in a different PR.

Reply:
> Thinking more about this my suggested solution is incorrect and I'm not against fixing Connect behaviour. Just to change the behaviour in a different PR.

SGTM. I can make that PR first.

Reply:
That would be amazing. Otherwise I can also take that up.

Reply:
It's probably going to be several patches. "transient" and "indirect" (through a proxy) are _not_ the same but the code seems to use these concepts interchangeably.

Reply:
> That would be amazing. Otherwise I can also take that up.

Ok, I'm going to be realistic... I'm not likely going to get to this soon (possibly within a month or two, but not "soon"). Go ahead and take it if you find time/motivation.